mail_edit/models/compose_message.py

```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+# © 2016 Sunflower IT (http://sunflowerweb.nl)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, fields, models
+
+
+class MailMessage(models.Model):
+    _inherit = "mail.message"
+
+    @api.model
+    def _get_model_selection(self):
+        """Get allowed models and their names."""
+        model_objs = self.env["res.request.link"].search(
+            [("mail_edit", "=", True)],
+            order="name")
+        return [(m.object, m.name) for m in model_objs]
+
+    @api.one
+    @api.onchange("destination_object_id")
+    def change_destination_object(self):
+        """Update some fields for the new message."""
+        # pylint: disable=api-one-deprecated
+        if self.destination_object_id:
+            self.model = self.destination_object_id._name
+            self.res_id = self.destination_object_id.id
+
+            model_name = self.env["ir.model"].search([
+                ("model", "=", self.model)]).name
+            display_name = self.destination_object_id.name_get()[0][1]
+            if model_name:
+                display_name = "%s %s" % (model_name, display_name)
+
+            self.record_name = display_name
+        else:
+            self.model = self.res_id = self.record_name = False
+
+    destination_object_id = fields.Reference(
+        _get_model_selection,
+        "Destination object",
+        required=True,
```

Comment:
don't make this required on the database level, if ever the constraint will be set up (it won't usually, because there are mails already in the system) things go bad. Set the flag on the view